File: node/private-hub-lock.ts
import { spawn, type ChildProcessWithoutNullStreams } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { getPrivateHelperPath } from './private-helper-paths';

export const PRIVATE_HUB_LOCK_FILE = '.private-hub.lock';
const HANDSHAKE_TIMEOUT_MS = 5_000;
const CLEANUP_TIMEOUT_MS = 5_000;
const cleanupFailures = new WeakSet<Error>();

/** Main-only identity evidence; messages and error codes cannot forge it. */
export function isPrivateHubLeaseCleanupFailure(error: unknown): error is Error {
  return error instanceof Error && cleanupFailures.has(error);
}

function cleanupFailure(): Error {
  const error = new Error('Private hub storage lease cleanup could not be confirmed.');
  cleanupFailures.add(error);
  return error;
}

async function confirmCleanup(operation: () => Promise<void>, timeout = CLEANUP_TIMEOUT_MS): Promise<void> {
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    await Promise.race([
      Promise.resolve().then(operation),
      new Promise<never>((_resolve, reject) => { timer = setTimeout(() => reject(cleanupFailure()), timeout); }),
    ]);
  } catch (error) {
    throw isPrivateHubLeaseCleanupFailure(error) ? error : cleanupFailure();
  } finally { clearTimeout(timer); }
}

export class PrivateHubLeaseError extends Error {
  readonly code = 'PRIVATE_HUB_LEASE_EXISTS';

  constructor() {
    super('This private hub already has an open storage session in another process.');
    this.name = 'PrivateHubLeaseError';
  }
}

function sameFile(left: fs.Stats, right: fs.Stats): boolean {
  return left.dev === right.dev && left.ino === right.ino;
}

async function canonicalDirectory(directory: string): Promise<fs.Stats> {
  const stat = await fs.promises.lstat(directory);
  if (!stat.isDirectory() || stat.isSymbolicLink() || await fs.promises.realpath(directory) !== directory) {
    throw new Error('Private hub leases require a canonical directory without symbolic links.');
  }
  return stat;
}

/**
 * Local-filesystem advisory lock on an open-file description shared by parent
 * and native helper. The parent opens with O_NOFOLLOW; the helper validates the
 * inherited descriptor against its directory-relative path and acquires flock.
 * Helper death invalidates the session but the parent retains the OS lock until
 * queued IO drains. Parent death closes its descriptor and the helper's stdin;
 * EOF closes the final reference. There is no stale PID/timeout reclamation.
 * The empty lock inode is persistent and NEVER unlinked. This arbitrates
 * cooperating processes, not hostile same-user processes or network storage.
 */
export class PrivateHubLease {
  readonly #directory: string;
  readonly #root: fs.Stats;
  readonly #child: ChildProcessWithoutNullStreams;
  readonly #handle: fs.promises.FileHandle;
  readonly #closed: Promise<void>;
  #owned: { dev: number; ino: number } | undefined;
  #lost = false;
  readonly #lossController = new AbortController();
  #releaseCompletion: Promise<void> | undefined;

  private constructor(directory: string, root: fs.Stats, child: ChildProcessWithoutNullStreams, handle: fs.promises.FileHandle) {
    this.#directory = directory;
    this.#root = root;
    this.#child = child;
    this.#handle = handle;
    this.#closed = new Promise(resolve => {
      child.once('close', () => {
        this.lose();
        resolve();
      });
    });
    child.once('exit', () => { this.lose(); });
    child.once('error', () => { this.lose(); });
    child.stdin.on('error', () => { this.lose(); });
  }

  static async acquire(directory: string): Promise<PrivateHubLease> {
    if (!['darwin', 'linux'].includes(process.platform)) {
      throw new Error('Private hub storage currently requires macOS or Linux advisory locks.');
    }
    const helper = getPrivateHelperPath('private-hub-lock');
    const root = await canonicalDirectory(directory);
    const handle = await fs.promises.open(path.join(directory, PRIVATE_HUB_LOCK_FILE),
      fs.constants.O_RDWR | fs.constants.O_CREAT | (fs.constants.O_NOFOLLOW || 0) | (fs.constants.O_NONBLOCK || 0), 0o600);
    let child: ChildProcessWithoutNullStreams;
    try {
      const file = await handle.stat();
      if (!file.isFile() || file.nlink !== 1 || file.size !== 0 || (file.mode & 0o077) !== 0
        || (process.geteuid && file.uid !== process.geteuid())
        || !sameFile(await canonicalDirectory(directory), root)) {
        throw new Error('Private hub storage locks require a private, empty regular file in the original directory.');
      }
      // FD 3 is a duplicate of this same open-file description. Keep the parent
      // FileHandle alive until release; helper death must not unlock pending IO.
      child = spawn(helper, [directory, String(root.dev), String(root.ino)], {
        cwd: directory,
        stdio: ['pipe', 'pipe', 'pipe', handle.fd],
      }) as ChildProcessWithoutNullStreams;
    } catch (error) {
      await confirmCleanup(() => handle.close());
      throw error;
    }
    const lease = new PrivateHubLease(directory, root, child, handle);
    try {
      const owned = await new Promise<{ dev: number; ino: number }>((resolve, reject) => {
        let output = '';
        let settled = false;
        const timer = setTimeout(() => finish(new Error('Private hub lock helper startup timed out.')), HANDSHAKE_TIMEOUT_MS);
        const finish = (error?: Error, value?: { dev: number; ino: number }) => {
          if (settled) {
            return;
          }
          settled = true;
          clearTimeout(timer);
          if (error) {
            reject(error);
          } else if (value) {
            resolve(value);
          }
        };
        child.once('error', () => finish(new Error('Private hub lock helper is unavailable. Build the privacy tools first.')));
        child.once('close', () => finish(new Error('Private hub lock helper exited before acquiring the lock.')));
        child.stderr.on('data', () => {
          // Never propagate arbitrary helper output into logs or UI.
          if (!settled) {
            finish(new Error('Private hub lock helper reported an unexpected error.'));
          }
          lease.lose();
          child.kill('SIGKILL');
        });
        child.stdout.on('data', (bytes: Buffer) => {
          if (settled) {
            lease.lose();
            child.kill('SIGKILL');
            return;
          }
          output += bytes.toString('ascii');
          if (output.length > 128) {
            finish(new Error('Invalid private hub lock helper response.'));
            return;
          }
          if (!output.includes('\n')) {
            return;
          }
          if (output === 'UNSUPPORTED_FS\n') {
            finish(new Error('Private hubs currently require a supported local filesystem; network volumes are not supported.'));
            return;
          }
          if (output === 'BUSY\n') {
            finish(new PrivateHubLeaseError());
            return;
          }
          const match = /^READY ([0-9]+) ([0-9]+)\n$/.exec(output);
          const dev = match ? Number(match[1]) : NaN;
          const ino = match ? Number(match[2]) : NaN;
          if (!Number.isSafeInteger(dev) || !Number.isSafeInteger(ino)) {
            finish(new Error('Private hub lock helper could not acquire a valid storage lock.'));
          } else {
            finish(undefined, { dev, ino });
          }
        });
      });
      lease.#owned = owned;
      await lease.assertOwned();
      return lease;
    } catch (error) {
      await lease.release();
      throw error;
    }
  }

  async assertOwned(): Promise<void> {
    this.assertLive();
    const root = await canonicalDirectory(this.#directory);
    if (!sameFile(root, this.#root)) {
      throw new Error('The private hub directory was replaced.');
    }
    const file = await fs.promises.lstat(path.join(this.#directory, PRIVATE_HUB_LOCK_FILE));
    const held = await this.#handle.stat();
    if (!sameFile(file, held) || !file.isFile() || file.isSymbolicLink() || file.nlink !== 1 || file.size !== 0
      || !this.#owned || file.dev !== this.#owned.dev || file.ino !== this.#owned.ino
      || (file.mode & 0o077) !== 0) {
      throw new Error('The private hub storage lock was replaced or changed.');
    }
    this.assertLive();
  }

  get lostSignal(): AbortSignal {
    return this.#lossController.signal;
  }

  private lose(): void {
    this.#lost = true;
    this.#lossController.abort();
  }

  private assertLive(): void {
    if (this.#lost || this.#releaseCompletion || this.#child.exitCode !== null || this.#child.signalCode !== null) {
      throw new Error('The private hub storage lock was lost; close and reopen the hub.');
    }
  }

  /** Called after queued IO drains. Concurrent releases share one completion. */
  release(): Promise<void> {
    if (!this.#releaseCompletion) {
      // Install the shared result before invoking external callbacks. A rejected
      // confirmation stays rejected even after the original work settles late.
      this.#releaseCompletion = Promise.resolve().then(async () => {
        let failure: Error | undefined;
        const stop = (): void => {
          try { this.#child.kill('SIGKILL'); } catch { /* Await close evidence below. */ }
        };
        const timer = setTimeout(stop, HANDSHAKE_TIMEOUT_MS);
        try {
          try { this.#child.stdin.end(); } catch { stop(); }
          // Allow graceful exit, then a bounded interval to confirm the kill.
          await confirmCleanup(() => this.#closed, HANDSHAKE_TIMEOUT_MS + CLEANUP_TIMEOUT_MS);
        } catch (error) {
          failure = isPrivateHubLeaseCleanupFailure(error) ? error : cleanupFailure();
        } finally { clearTimeout(timer); }
        try {
          // Never use LOCK_UN: the two duplicated descriptors share one flock.
          // Attempt our final close even when the helper's close is unconfirmed.
          await confirmCleanup(() => this.#handle.close());
        } catch (error) {
          failure ??= isPrivateHubLeaseCleanupFailure(error) ? error : cleanupFailure();
        } finally { this.lose(); }
        if (failure) { throw failure; }
      });
      void this.#releaseCompletion.catch(() => undefined);
    }
    return this.#releaseCompletion;
  }
}

File: bin/private-hub-lock.c
#define _DARWIN_C_SOURCE
#define _XOPEN_SOURCE 700
#define _DEFAULT_SOURCE
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#ifdef __APPLE__
#include <sys/mount.h>
#elif defined(__linux__)
#include <sys/vfs.h>
#include <linux/magic.h>
#endif

static int local_filesystem(int directory) {
  struct statfs info;
  if (fstatfs(directory, &info) != 0) return 0;
#ifdef __APPLE__
  return (info.f_flags & MNT_LOCAL) != 0;
#elif defined(__linux__)
  /* Deliberately conservative: unlisted/FUSE/network filesystems are rejected.
   * ext2/3/4 share EXT2_SUPER_MAGIC. More filesystems require separate testing. */
  return info.f_type == EXT2_SUPER_MAGIC || info.f_type == XFS_SUPER_MAGIC
    || info.f_type == BTRFS_SUPER_MAGIC || info.f_type == TMPFS_MAGIC
    || info.f_type == OVERLAYFS_SUPER_MAGIC;
#else
  (void)directory;
  return 0;
#endif
}

/* Never unlink the lock: all contenders must flock the same persistent inode.
 * No passwords, keys or media are passed to this helper. Parent death closes
 * stdin and its shared descriptor; EOF closes the helper's final reference.
 * Helper death alone leaves the parent's shared lock held until its IO drains.
 * Never call LOCK_UN: that would unlock both shared references prematurely. */
static int same_file(const struct stat *a, const struct stat *b) {
  return a->st_dev == b->st_dev && a->st_ino == b->st_ino;
}

static int fail(void) {
  puts("ERROR");
  fflush(stdout);
  return 1;
}

int main(int argc, char **argv) {
  if (argc != 4) return fail();
  char *canonical = realpath(argv[1], NULL);
  if (!canonical || strcmp(canonical, argv[1]) != 0) {
    free(canonical);
    return fail();
  }
  free(canonical);
  char *end_dev = NULL, *end_ino = NULL;
  errno = 0;
  uintmax_t expected_dev = strtoumax(argv[2], &end_dev, 10);
  uintmax_t expected_ino = strtoumax(argv[3], &end_ino, 10);
  if (errno || !end_dev || *end_dev || !end_ino || *end_ino) return fail();
  int directory = open(argv[1], O_RDONLY | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
  struct stat root, current, owned;
  if (directory < 0 || fstat(directory, &root) != 0 || !S_ISDIR(root.st_mode)
      || (uintmax_t)root.st_dev != expected_dev || (uintmax_t)root.st_ino != expected_ino) return fail();
  if (!local_filesystem(directory)) {
    puts("UNSUPPORTED_FS");
    fflush(stdout);
    return 3;
  }
  const int lock = 3; /* Inherited duplicate of the parent's retained handle. */
  if (fcntl(lock, F_SETFD, FD_CLOEXEC) != 0 || fstat(lock, &owned) != 0 || !S_ISREG(owned.st_mode)
      || owned.st_nlink != 1 || owned.st_size != 0 || owned.st_uid != geteuid()
      || (owned.st_mode & 077) != 0) return fail();
  if (flock(lock, LOCK_EX | LOCK_NB) != 0) {
    if (errno == EWOULDBLOCK || errno == EAGAIN) {
      puts("BUSY");
      fflush(stdout);
      return 2;
    }
    return fail();
  }
  if (fstatat(directory, ".private-hub.lock", &current, AT_SYMLINK_NOFOLLOW) != 0
      || !same_file(&current, &owned) || current.st_nlink != 1 || !S_ISREG(current.st_mode)
      || lstat(argv[1], &current) != 0 || !same_file(&root, &current) || !S_ISDIR(current.st_mode)) return fail();
  if (printf("READY %ju %ju\n", (uintmax_t)owned.st_dev, (uintmax_t)owned.st_ino) < 0
      || fflush(stdout) != 0) return 1;
  char byte;
  ssize_t count;
  do {
    count = read(STDIN_FILENO, &byte, 1);
  } while (count < 0 && errno == EINTR);
  /* Only EOF is a normal helper exit. The parent retains its shared lock
   * reference until queued IO drains, including on unexpected input. */
  close(lock);
  close(directory);
  return count == 0 ? 0 : 1;
}
